src/optimizer/cost_and_stats_calculator.cpp

```diff
@@ -10,127 +10,306 @@
+			std::shared_ptr<GroupExpression> gexpr,
+			const PropertySet *output_properties,
+			const std::vector<PropertySet> *input_properties_list,
+			std::vector<std::shared_ptr<Stats>> child_stats,
+			std::vector<double> child_costs) {
+			gexpr_ = gexpr;
+			output_properties_ = output_properties;
+			input_properties_list_ = input_properties_list;
+			child_stats_ = child_stats;
+			child_costs_ = child_costs;
+
+			gexpr->Op().Accept(this);
+		}
+
+		void CostAndStatsCalculator::Visit(const DummyScan *) {
+			// TODO: Replace with more accurate cost
+			output_stats_.reset(new Stats(nullptr));
+			output_cost_ = 0;
+		}
+
+		void CostAndStatsCalculator::Visit(const PhysicalSeqScan * op) {
+			// TODO : Replace with more accurate cost
+			// retrieve table_stats from catalog by db_id and table_id
+			auto stats_storage = StatsStorage::GetInstance();
+			auto table_stats = std::dynamic_pointer_cast<TableStats>(
+				stats_storage->GetTableStats(op->table_->GetDatabaseOid(), op->table_->GetOid()));
+      auto property_ = output_properties_->GetPropertyOfType(PropertyType::COLUMNS)->As<PropertyColumns>();
+			auto output_stats = generateOutputStat(table_stats, property_);
+
+			auto predicate_prop =
+				output_properties_->GetPropertyOfType(PropertyType::PREDICATE)
+					->As<PropertyPredicate>();
+			if (predicate_prop == nullptr) {
+				output_cost_ = Cost::NoConditionSeqScanCost(table_stats);
+				output_stats_.reset(output_stats.get());
+				return;
+			}
+
+			std::vector<oid_t> key_column_ids;
+			std::vector<ExpressionType> expr_types;
+			std::vector<type::Value> values;
+
+
+			expression::AbstractExpression *predicate = predicate_prop->GetPredicate();
+			bool canSearch;
+
+			// traverse the predicate and retrieve the several tuple value expressions
+			util::GetPredicateColumns(op->table_->GetSchema(), predicate,
+													key_column_ids, expr_types,
+													values, canSearch);
+
+			if (key_column_ids.size() == 1) { // single condition
+				ValueCondition condition(key_column_ids.at(0), expr_types.at(0), values.at(0));
+				output_cost_ = Cost::SingleConditionSeqScanCost(table_stats, condition, output_stats);
+
+			} else if (key_column_ids.size() == 2 &&
+								 (predicate->GetExpressionType() == ExpressionType::CONJUNCTION_AND ||
+									predicate->GetExpressionType() == ExpressionType::CONJUNCTION_OR)){ //conjunction expression
+				ValueCondition lhs_cond(key_column_ids.at(0), expr_types.at(0), values.at(0));
+				ValueCondition rhs_cond(key_column_ids.at(1), expr_types.at(1), values.at(1));
+				//Get the scan_cost
+				output_cost_ = Cost::NoConditionSeqScanCost(table_stats);
+				//Update output stats num_rows
+				updateConjunctionStats(table_stats, lhs_cond, rhs_cond, predicate->GetExpressionType(), output_stats);
+
+			} else {
+
+				// TODO: support multiple conjunctions
+				output_cost_ = 2;
+			}
+
+		};
+
+		void CostAndStatsCalculator::Visit(const PhysicalIndexScan *op) {
```

Comment:
Seems like the code here is somewhat repetitive. Can we refactor some of the logic out? and make those two scans share the same logic?